lib/ecto/association.ex

```diff
@@ -524,6 +529,17 @@ defmodule Ecto.Association.Has do
   end
 
   @doc false
+  def preload(%{related_key: related_key, owner_key: owner_key} = refl, repo, query, fields, prefix, structs, opts) do
+    ids = structs |> Enum.map(&(Map.get(&1, owner_key)))
+    query = assoc_query(refl, query, Enum.uniq(ids))
+    Ecto.Repo.Preloader.fetch_query(structs, refl, query, prefix, {0, related_key}, fields)
+    |> case do
+      list when is_list(list) -> list
+      query -> repo.all(query, opts)
+    end
+  end
+
+  @doc false
   def preload_info(%{related_key: related_key} = refl) do
```

Comment:
I wasnt sure if you intended me to remove this callback but it does seem to be used by other parts of the code... should this be removed with this change or is it still valuable?